
*   Nama     : La Liani
*   Program  : FTDS18-rmt
*   Tugas    : GC3










### Assignment Problems
 The Look yang merupakan salah satu platform e-commerce terbesar di planet Mars. sehingga perlu untuk membuat laporan evaluasi aktivitas penjualan di platform tersebut.

Untuk mempermudah proses evaluasi yang terarah, maka kita harus menentukan problem statement dengan success, criteria berdasarkan SMART. Untuk itu, kita perlu menentukan problem statement berdasarkan penjabaran-penjabaran analisis dalam bentuk persoalan yang akan  dijawab menggunakan Query SQL. (SMART akan berdasarkan dari poin-poin penjabaran).



### Problem Statement

Untuk melakukan evaluasi hasil transaksi penjualan menjadi lebih terarah, maka kita perlu tentukan terlebih dahulu problem statement yang akan menjadi poin evaluasi dari transaksi penjualan dari e-commerce tersebut. Ada beberapa poin yang akan kita jadikan poin evaluasi sebagai berikut:


1. Berapa jumlah transaksi yang berstatus Complete tiap bulan selama Q1 sampai Q3 di tahun 2022? Insight apa yang bisa kamu berikan?
2. Berikan informasi total penjualan (dalam USD) tiap bulan selama Q1 sampai Q3 di tahun 2022! (Hanya yang transaksi berstatus Complete. Apa informasi yang bisa kamu sampaikan?
3. Berapa user yang melakukan transaksi berstatus Complete di tiap bulan dari Q1 sampai Q3 2022? Apa kesimpulanmu?
4. Kategori produk apa saja yang paling banyak dibeli (status transaksi: Complete) di tiap bulannya selama Q1 sampai Q3 tahun 2022? Berikan insight!
5. Kategori produk apa saja yang paling banyak mendapatkan pendapatan (staus transaksi: Complete) di tiap bulannya selama Q1 sampai Q3 tahun 2022? Berikan insight!
6. Dibandingkan dengan jumlah transaksi dan total penjualan, mana yang paling berkaitan dengan jumlah user yang melakukan transaksi? Apa analisis yang dapat kamu jelaskan? (Hint: Kamu bisa menggunakan korelasi)
PERHATIAN!. Untuk semua penjabaran tidak boleh menggunakan bantuan Pandas Query atau method lainnya untuk menyeleksi, menggabungkan, memanipulasi data. HANYA gunakan Query SQL. Tidak perlu menampilkan grafik, hanya berupa dataframe.

### Instal, Import package 

In [ ]:
! pip install google-cloud-bigquery -q

In [1]:
from google.colab import auth
from google.cloud import bigquery
import pandas as pd



### Membuat Koneksi ke Bigquery Public Data

In [2]:
auth.authenticate_user()
project_id = "refined-asset-376308"
client = bigquery.Client(project= project_id )

### Load data dari Bigquery Public Data

In [55]:
# membuat koneksi ke bigquery sekaligus meload data
df1 = client.query("""
SELECT * FROM `bigquery-public-data.thelook_ecommerce.order_items` LIMIT 1000

""").to_dataframe()
df1

id  order_id  user_id  product_id  inventory_item_id     status  \
0     75076     51556    41381       13606             203319    Shipped   
1    118747     81570    65442       13606             321141    Shipped   
2    151249    104163    83551       13606             408707   Complete   
3    171188    117911    94564       13606             462691   Complete   
4     24820     17024    13616       13606              67128  Cancelled   
..      ...       ...      ...         ...                ...        ...   
995  104969     72125    57973       11114             283919    Shipped   
996  106610     73224    58821       18218             288359    Shipped   
997  107871     74106    59522       13726             291729    Shipped   
998  110981     76262    61266       11114             300085    Shipped   
999  111097     76342    61328       24554             300400    Shipped   

                          created_at                       shipped_at  \
0          2022-11-28 07:41:06+00:00        2022-11-25 02:28:00+00:00   
1   2023-02-18 05:33:36.794320+00:00 2023-02-19 09:57:20.794320+00:00   
2          2020-05-24 07:52:03+00:00        2020-05-27 05:57:00+00:00   
3          2022-03-21 10:50:31+00:00        2022-03-20 15:46:00+00:00   
4          2022-07-06 22:06:32+00:00                              NaT   
..                               ...                              ...   
995        2023-01-17 14:05:22+00:00        2023-01-18 18:04:00+00:00   
996        2022-09-08 14:40:08+00:00        2022-09-08 20:25:00+00:00   
997        2022-07-13 23:47:07+00:00        2022-07-16 05:05:00+00:00   
998        2021-10-27 04:54:00+00:00        2021-10-30 00:37:00+00:00   
999        2022-11-08 09:43:14+00:00        2022-11-09 14:40:00+00:00   

                 delivered_at returned_at  sale_price  
0                         NaT         NaT         2.5  
1                         NaT         NaT         2.5  
2   2020-05-29 12:46:00+00:00         NaT         2.5  
3   2022-03-23 09:12:00+00:00         NaT         2.5  
4                         NaT         NaT         2.5  
..                        ...         ...         ...  
995                       NaT         NaT         9.0  
996                       NaT         NaT         9.0  
997                       NaT         NaT         9.0  
998                       NaT         NaT         9.0  
999                       NaT         NaT         9.0  

[1000 rows x 11 columns]

  Untuk mengetahui jumlah transaksi dengan status complete setiap bulan selama Q1 hingga Q3 pada tahun 2022 adalah terlebih dahulu mendefenisikan Q1 hingga Q3 adalah dari januari sampai september. untuk transaksi itu sendiri adalah banyaknya order_id yang statusnya complete. untuk menyatakan status complete digunakan kolom delivered_at karena jika menggunakan yang lain seperti created_at itu masih ada kemungkinan untuk NaT. dan perlu untuk diorder by terlebih dahulu. sehingga ditampilkan tiga hasil query dan hasilnya berbeda yang kemungkinan karena ada NaT yang masuk sehingga ikut terhitung sebagai transaksi complete atau karena perbedaan waktu kirim misalnya created_at hari ini tapi delivered_at itu dua hari setelahnya sehinnga itu tidak masuk sebagai range yang ditentukan yakni dari januari sampai september. disini kita menggunakan Subquery, dan insight yang bisa diberikan kita akan lebih cenderung memilih jumlah transaksi berdasarkan kolom delivered_at. hal ini diperkuat dari hasil jumlah transaksinya yang paling kecil yakni 13685. artinya meskipun transaksinya sudah created_at dan delivered_at tapi blum tentu bisa delivered_at.

In [61]:
df2 = client.query("""
SELECT COUNT(order_id) AS jumlah_order
FROM
(SELECT 
   order_id AS order_id,
   created_at AS order_time,
   delivered_at AS send_time,
   shipped_at   AS shipped_time,
       
FROM `bigquery-public-data.thelook_ecommerce.order_items` 

WHERE status = "Complete" 
      AND created_at >= "2022-01-01"
ORDER BY created_at DESC)

WHERE order_time <= "2022-09-30"
      
""").to_dataframe()
df2

jumlah_order
0         13772

diatas adalah jumlah transaksi berdasakan kolom created_at

In [60]:
df3 = client.query("""
SELECT COUNT(order_id) AS jumlah_order
FROM
(SELECT 
   order_id AS order_id,
   created_at AS order_time,
   delivered_at AS send_time,
   shipped_at   AS shipped_time,
       
FROM `bigquery-public-data.thelook_ecommerce.order_items` 

WHERE status = "Complete" 
      AND delivered_at >= "2022-01-01"
ORDER BY delivered_at DESC)

WHERE send_time <= "2022-09-30"
      
""").to_dataframe()
df3

jumlah_order
0         13685

diatas adalah jumlah transaksi berdasakan kolom delivered_at

In [62]:
df2 = client.query(""" # menggunakan subquey untuk dapat menentukan jumlah orderan dengan status complete
SELECT COUNT(order_id) AS jumlah_order
FROM
(SELECT 
   order_id AS order_id,
   created_at AS order_time,
   delivered_at AS send_time,
   shipped_at   AS shipped_time,
       
FROM `bigquery-public-data.thelook_ecommerce.order_items` 

WHERE status = "Complete" 
      AND shipped_at >= "2022-01-01"
ORDER BY shipped_at DESC)

WHERE shipped_time <= "2022-09-30"
      
""").to_dataframe()
df2

jumlah_order
0         13758

diatas adalahjumlah transaksi berdasakan kolom created_at

-Untuk mengetahui total penjualan maka kita akan menghitung total sale_price dengan status complete dari Q1 hingga Q3 selma tahun 2022 berdasarkan kolom delivered_at.

In [64]:
df3 = client.query(""" # sama seperti query sebelumnya tapi kolom yang menjadi perhitungan adalah kolom sale_price dengan status complete
SELECT SUM(total_uang) AS Total_uang
FROM
(SELECT 
   sale_price AS total_uang,
   created_at AS order_time,
   delivered_at AS send_time,
   shipped_at   AS shipped_time,
       
FROM `bigquery-public-data.thelook_ecommerce.order_items` 

WHERE status = "Complete" 
      AND delivered_at >= "2022-01-01"
ORDER BY delivered_at DESC)

WHERE send_time <= "2022-09-30"
      
""").to_dataframe()
df3

Total_uang
0  799541.180573

dari sini kita bisa mengetahui bahwa total hasil transaksi(USD) yang bersifat complete adalah sebesar $799541.180573

- Untuk mengetahui total user yang melakukan transaksi dengan status transaksi yang bersifat complete maka kita akan menghitung jumlah user_id dengan status complete.

In [69]:
df3 = client.query(""" # sama seperti query sebelumnya tapi kolom yang menjadi perhitungan adalah kolom user_id dengan status complete
SELECT COUNT(jumlah_pembeli) AS jumlah_pembeli
FROM
(SELECT 
   user_id AS jumlah_pembeli,
   created_at AS order_time,
   delivered_at AS send_time,
   shipped_at   AS shipped_time,
       
FROM `bigquery-public-data.thelook_ecommerce.order_items` 

WHERE status = "Complete" 
      AND delivered_at >= "2022-01-01"
ORDER BY delivered_at DESC)

WHERE send_time <= "2022-09-30"
      
""").to_dataframe()
df3

jumlah_pembeli
0           13685

dapat diketahui jumlah user yang melakukan transaksi dengan status complete sebanyak 13685

menggunakan clausa COUNT DISTINCT untuk cek apakah ada user_id yang berulang.

In [70]:
df3 = client.query(""" # disini kita menggunakan clausa DISTINCT untuk mengetahui apakah ada pembeli yang melakukan pemebelian berulang dan hasilnya ada.
                       #sama seperti query sebelumnya tapi kolom yang menjadi perhitungan adalah kolom user_id dengan status complete
SELECT COUNT(DISTINCT jumlah_pembeli) AS jumlah_pembeli
FROM
(SELECT 
   user_id AS jumlah_pembeli,
   created_at AS order_time,
   delivered_at AS send_time,
   shipped_at   AS shipped_time,
       
FROM `bigquery-public-data.thelook_ecommerce.order_items` 

WHERE status = "Complete" 
      AND delivered_at >= "2022-01-01"
ORDER BY delivered_at DESC)

WHERE send_time <= "2022-09-30"
      
""").to_dataframe()
df3

jumlah_pembeli
0            8840

total pembeli yang berbeda adalah sebanyak 8840.

 - Load data products

In [89]:

df1 = client.query("""
SELECT 
   product_id,
FROM `bigquery-public-data.thelook_ecommerce.order_items`
WHERE
   status = "Complete"
""").to_dataframe()
df1

product_id
0           13606
1           13606
2           28951
3           28557
4           28557
...           ...
44901         146
44902         146
44903         146
44904        3449
44905        3449

[44906 rows x 1 columns]

melakukan penggabungan dua tabel product dengan order_items

In [92]:
df1 = client.query("""
SELECT 
   product_id,
FROM `bigquery-public-data.thelook_ecommerce.order_items`
WHERE
   status = "Complete"
INNER JOIN product_id ON product_id = product_id
(SELECT id AS product_id, retail_price, category FROM `bigquery-public-data.thelook_ecommerce.products`)

""").to_dataframe()
df1

BadRequest: ignored

In [ ]:
df1 = client.query("""
 SELECT *   FROM bigquery-public-data.thelook_ecommerce.distribution_centers LIMIT 1000
  
""").to_dataframe()
df1

id                                         name  latitude  longitude
0   1                                   Memphis TN   35.1174   -89.9711
1   2                                   Chicago IL   41.8369   -87.6847
2   3                                   Houston TX   29.7604   -95.3698
3   4                               Los Angeles CA   34.0500  -118.2500
4   5                               New Orleans LA   29.9500   -90.0667
5   6  Port Authority of New York/New Jersey NY/NJ   40.6340   -73.7834
6   7                              Philadelphia PA   39.9500   -75.1667
7   8                                    Mobile AL   30.6944   -88.0431
8   9                                Charleston SC   32.7833   -79.9333
9  10                                  Savannah GA   32.0167   -81.1167